New Rule: newline-before-return

artursvonda opened the pull request:
#### PR checklist

- [ ] Addresses an existing issue:
- [x] New feature, bugfix, or enhancement
- [x] Includes tests
- [ ] Documentation update

#### What changes did you make?

A new rule to force blank lines before return. Based on http://eslint.org/docs/rules/newline-before-return

#### Is there anything you'd like reviewers to focus on?

Making sure all test cases are covered. I took the base file from eslint and added a few cases.

palantirtech commented:
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @artursvonda! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

ajafff commented on src/rules/newlineBeforeReturnRule.ts:
```diff
@@ -0,0 +1,87 @@
+
+    private getIndex(node: ts.Node) {
+        const children = node.parent ? node.parent.getChildren() : [];
+        // For some reason instances are not the same and indexOf doesn't work with given node directly
```
`getChildren` does not return what you expect, that's why `indexOf` does not work here:

`node.parent` is either `ts.BlockLike` or `ts.IfStatement`, `ts.IterationStatement`, ...
If it is *not* `ts.BlockLike` the return statement is the only child, e.g. an if statement without a block and only a single return statement. So you don't have to call `getChildren`.
If it is `ts.BlockLike` `getChildren` will return `[OpenBraceToken, SyntaxList, CloseBraceToken]` where SyntaxList is an array of statements within the block. But you don't need this SyntaxList, because BlockLike has a `statements` property (which also contains all statements, duh)

here is a simpler solution:
```ts
private getIndex(node: ts.Node) {
  const parent = node.parent!; // use ! (non null assertion) because we know every node has a parent (except ts.SourceFile)
  if (isBlockLike(parent) {
    return parent.statements.indexOf(node);
  } {
    // `node` is the only statement within this "block scope" -> do whatever you need to do in this case
  }
}
...
function isBlockLike(node: ts.Node): node is ts.BlockLike {
    return "statements" in node; // could also be node.kind === ts.SyntaxKind.Block || node.kind === ts.SyntaxKind.CaseClause || to be continued ...
}
```

ajafff commented on src/rules/newlineBeforeReturnRule.ts:
```diff
@@ -0,0 +1,87 @@
+        const position = source.getPositionOfLineAndCharacter(line - 1, 0);
+        const end = source.getPositionOfLineAndCharacter(line, 0) - 1;
+        const all = source.getText().substr(position, end - position);
+        const trimmed = all.replace(/\s+/, "");
```
you could also just use `all.search(/\S/) !== -1`

ajafff commented on src/rules/newlineBeforeReturnRule.ts:
```diff
@@ -0,0 +1,87 @@
+        super.visitReturnStatement(node);
+
+        let start = node.getStart();
+        const comments = ts.getLeadingCommentRanges(node.getSourceFile().text, node.getFullStart());
```
use `this.getSourceFile()` instead of `node.getSourceFile()`
the latter walks up the whole AST until it finds the SourceFile, which is unnecessary work

artursvonda commented:
Thanks @ajafff for all the comments and suggestions. I addressed those comments and simplified the code a bit.

nchen63 commented on src/rules/newlineBeforeReturnRule.ts:
```diff
@@ -0,0 +1,87 @@
+
+    private validateBlankLine(line: number) {
+        const source = this.getSourceFile();
+        const position = source.getPositionOfLineAndCharacter(line - 1, 0);
```
what would happen there was a return on the first line?

does this handle other cases where some writes a return statement on the same line as another statement?

artursvonda replied:
Good point. Not sure how that should be handled. Any ideas/suggestions? Meanwhile I'll add a test case.

nchen63 commented:
I'm leaning toward always flagging if there's another statement on the same line as the return statement

artursvonda commented:
@nchen63 Yeah, makes sense. The idea is that `return` should have spacing around it so it's clear and easier to spot and read. So the rule basically is "if the `return` is not the first thing in the block, there should be blank line before it.". With that in mind I simplified the code a bit more. Also changed the position of the error a bit and now it's at the start of the return statement.

nchen63 commented:
@artursvonda thanks!

nchen63 commented:
@artursvonda this rule is a good candidate for an auto fixer too

artursvonda commented:
@nchen63 No problem. And yeah, will see if I can add the fixed over some evening.

glen-84 commented:
This should probably have been called `blank-line-before-return`.
